src/lib/Occurrence.hs

```diff
@@ -0,0 +1,671 @@
+--     \ref.
+--       compile part1 ref
+--       compile part2 ref
+--       ...
+--
+-- - `ForEach b sub` is iteration
+--     \ref. for b in (type of b).
+--       compile sub ref
+--   The type that `b` iterates over can in principle be fetched from
+--   the Dex program that led to this `ForEach` constructor.  It only
+--   matters for the semantics, so we don't actually store it.
+
+-- === Use ===
+
+-- The Use structure is a "data space" representation of accessing
+-- arrays.
+--
+-- A Use implicitly iterates over as many dimensions of the accessed
+-- object as there are nesting levels of At constructors in the Use.
+-- We do not use explicit binders for this because they always appear
+-- in the same nesting order, so it's simplest to leave them implicit.
+-- For example, the case
+--   xs = ...
+--   for i. for j. xs.i.j
+-- is represented by the Use
+--   At IxAll (At IxAll One)
+-- meaning "we access every element or subarray once, two dimensions
+-- deep".
+--
+-- Note that a Use can still depend on iteration-space indices (from
+-- the `iter` scope)---it just doesn't have internal binders that are
+-- meant to be interpreted in iteration space.  For example, the Use
+-- for `xs` in
+--   for i.
+--     xs = ...
+--     xs.i
+-- is
+--   At (Var i) One
+-- meaning that `xs` is accessed at the location `i`.
+
+data Use (iter::S) =
+  -- The object is accessed in its entirety
+    Const Count
+  -- The object is accessed at the positions given by the first
+  -- argument.  The subarray at those positions is accessed according
+  -- to the sub-Use.  Note that at this point we permit the `IxAll`
+  -- constructor in `IxExpr`, so an `IxExpr` may represent a set of
+  -- positions.
+  | At (IxExpr iter) (Use iter)
+  -- The object is accessed in at most one of the given Use patterns.
+  | Max [Use iter]
+  -- The object is accessed in all of the given Use patterns.
+  | Sum [Use iter]
+  deriving (Eq, Show)
+
+-- The semantics of Use are similar to those for Access.  Here is the
+-- compilation that turns a Use into a program that dynamically counts
+-- how many times each index is used.
+--
+-- Note that now IxExpr permits the `IxAll` constructor, so an
+-- `IxExpr` represents a subset of indices rather than a single index.
+--
+-- - `Const ct` becomes accumulation
+--     \ref. ref += ct
+--     Also polymorphic in the ref type, applying elementwise
+--
+-- - `Max alts` becomes case
+--     \ref. case randint (length alts) of
+--       1 -> compile alt1 ref
+--       2 -> compile alt2 ref
+--       ...
+--
+-- - `Sum parts` becomes sequencing
+--     \ref.
+--       compile alt1 ref
+--       compile alt2 ref
+--       ...
+--
+-- - `At expr use` becomes testing and slicing
+--     \ref. for b in <next object dim>.
+--       if b in expr then (compile use ref!b)
+
+-- This semantics implies a transformation that we perform eagerly:
+-- - It's safe to distribute Sum into Max, so that Max ends up on the
+--   outside.
```

Comment:
With `Sum` it is safe, because we take the combinatorial hit to actually distribute it.  That is,
`max(a, b) + max(c, d) <= max(a + c, a + d, b + c, b + d)`.  In the `for` case we were doing the equivalent of turning `max(a, b) + max(c, d)` into `max(a + c, b + d)`, which indeed is not safe.

Reply:
Ok, I agree that if you take a combinatorial hit then it is indeed safe, but that's not what I would understand from this comment.